ibis/file/client.py

```diff
@@ -0,0 +1,137 @@
+import ibis
+import ibis.expr.types as ir
+from ibis.pandas.core import execute
+
+try:
+    import pathlib
+except ImportError:
+
+    # py2 compat
+    import pathlib2 as pathlib
+
+
+class FileClient(ibis.client.Client):
+
+    def __init__(self, root):
+        self.root = pathlib.Path(str(root))
+        self.dictionary = {}
+
+    def insert(self, path, expr, **kwargs):
+        raise NotImplementedError
+
+    def table(self, name, path):
+        raise NotImplementedError
+
+    def database(self, name=None, path=None):
+        if name is None:
+            return FileDatabase('root', self, path=path)
+
+        if name not in self.list_databases(path):
+            raise AttributeError(name)
+        if path is None:
+            path = self.root
+
+        new_name = "{}.{}".format(name, self.extension)
+        if (self.root / name).is_dir():
+            path /= name
+        elif not str(path).endswith(new_name):
+            path /= new_name
+
+        return FileDatabase(name, self, path=path)
+
+    def execute(self, expr, params=None, **kwargs):  # noqa
+        assert isinstance(expr, ir.Expr)
+        scope = kwargs.pop('scope', {})
+        return execute(
+            expr, scope=scope,
+            params=params, **kwargs)
+
+    def list_tables(self, path=None):
+        raise NotImplementedError
+
+    def _list_tables_files(self, path=None):
+        # tables are files in a dir
+        if path is None:
+            path = self.root
+
+        tables = []
+        if path.is_dir():
+            for d in path.iterdir():
+                if d.is_file():
+                    if str(d).endswith(self.extension):
+                        tables.append(d.stem)
+        elif path.is_file():
+            if str(path).endswith(self.extension):
+                tables.append(path.stem)
+        return tables
+
+    def list_databases(self, path=None):
+        raise NotImplementedError
+
+    def _list_databases_dirs(self, path=None):
+        # databases are dir
+        if path is None:
+            path = self.root
+
+        tables = []
+        if path.is_dir():
+            for d in path.iterdir():
+                if d.is_dir():
+                    tables.append(d.name)
+        return tables
+
+    def _list_databases_dirs_or_files(self, path=None):
+        # databases are dir & file
+        if path is None:
+            path = self.root
+
+        tables = []
+        if path.is_dir():
+            for d in path.iterdir():
+                if d.is_dir():
+                    tables.append(d.name)
+                elif d.is_file():
+                    if str(d).endswith(self.extension):
+                        tables.append(d.stem)
+        elif path.is_file():
+            # by definition we are at the db level at this point
```

Comment:
Why is this true?

Reply:
for HDF5 a file is a database, as well as a dir.

Reply:
Should this be in the HDF client then?

Reply:
no, if you look above you requested that the common code be moved here.

Reply:
Right, I definitely remember that request, but this method is only used in one place and makes assumptions that are specific to HDF5 which is why I'm asking. This isn't a blocker though.